datalake/src/main/java/com/sequenceiq/datalake/events/DatalakeStructuredFlowEventFactory.java

```diff
@@ -0,0 +1,100 @@
+ * The Event Handler has a Spring injected dependency on {@code CDPStructuredFlowEventFactor}, which is the interface of this class.
+ */
+@Component
+public class DatalakeStructuredFlowEventFactory implements CDPStructuredFlowEventFactory {
+
+    @Inject
+    private Clock clock;
+
+    @Inject
+    private NodeConfig nodeConfig;
+
+    @Inject
+    private SdxService sdxService;
+
+    @Inject
+    private SdxClusterDtoConverter sdxClusterDtoConverter;
+
+    @Inject
+    private SdxStatusRepository sdxStatusRepository;
+
+    @Value("${info.app.version:}")
+    private String serviceVersion;
+
+    @Override
+    public CDPStructuredFlowEvent<SdxClusterDto> createStructuredFlowEvent(Long resourceId, FlowDetails flowDetails, Boolean detailed) {
+        return createStructuredFlowEvent(resourceId, flowDetails, detailed, null);
+    }
+
+    @Override
+    public CDPStructuredFlowEvent<SdxClusterDto> createStructuredFlowEvent(Long resourceId, FlowDetails flowDetails, Boolean detailed, Exception exception) {
+        SdxCluster sdxCluster = sdxService.getById(resourceId);
+        CDPOperationDetails operationDetails = makeCdpOperationDetails(resourceId, sdxCluster);
+        SdxClusterDto sdxClusterDto = sdxClusterDtoConverter.sdxClusterToDto(sdxCluster);
+        SdxStatusEntity sdxStatus = sdxStatusRepository.findFirstByDatalakeIsOrderByIdDesc(sdxCluster);
+        String status = sdxStatus.getStatus().name();
+        String statusReason = sdxStatus.getStatusReason();
+        CDPStructuredFlowEvent<SdxClusterDto> event = new CDPStructuredFlowEvent<>(operationDetails, flowDetails, sdxClusterDto, status, statusReason);
+        if (exception != null) {
+            event.setException(ExceptionUtils.getStackTrace(exception));
+        }
+        return event;
+    }
+
+    private CDPOperationDetails makeCdpOperationDetails(Long resourceId, SdxCluster sdxCluster) {
+        // turning the string constants in CloudbreakEventService into a enums would be nice.
+        String resourceType = CloudbreakEventService.DATALAKE_RESOURCE_TYPE;
+
+        // todo: make a CDPOperationDetails Builder
+        CDPOperationDetails operationDetails = new CDPOperationDetails();
+        operationDetails.setTimestamp(clock.getCurrentTimeMillis());
+        operationDetails.setEventType(FLOW);
+        operationDetails.setResourceId(resourceId);
+        operationDetails.setResourceName(sdxCluster.getName());
+        operationDetails.setResourceType(resourceType);
+        operationDetails.setCloudbreakId(nodeConfig.getId());
+        operationDetails.setCloudbreakVersion(serviceVersion);
+        operationDetails.setResourceCrn(sdxCluster.getResourceCrn());
+        operationDetails.setUserCrn(ThreadBasedUserCrnProvider.getUserCrn());
+        operationDetails.setAccountId(sdxCluster.getAccountId());
+        operationDetails.setEnvironmentCrn(sdxCluster.getEnvCrn());
+        operationDetails.setResourceEvent(ResourceEvent.DATALAKE_DATABASE_BACKUP.name());
```

Comment:
You don't need to define here, because we store in FlowDetails the current and next event

Reply:
I've removed this line.